Issue 15: coding guidelines

LinhWuerzburger opened the pull request:
First attempt to close Issue #15, defining coding guidelines. At least the things which are obvious at the moment.

LinhWuerzburger commented:
do you think we should add something like: if you want to contribute, pls open a PR?

LinhWuerzburger commented:
I changed the name from contributing to coding guidelines because it only has information on the coding guidelines and nothing more. a guide for contributing is in the wiki. the question is: do we want to link to the wiki or not? Or should we make it the other way around and link in the wiki to a markdown file?

ChristianVonMach reviewed (changes requested):
There are some other points in the Googles style guide, that needs to bee addressed. I don't think we should handle all at onces. So this are the thirst few points.

LinhWuerzburger commented on CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,52 @@
+## Contributing Guidelines
+
+The coding guidelines of ARTSS are mainly based on the [google coding guidelines](https://google.github.io/styleguide/cppguide.html). Below there are some rules which are not defined in the google standard or we want to especially emphasise them in order to achieve the most consistent code possible.
```
We don't use C++20 yet see #105 
Regarding the second point, I don't know what you mean with non-standard extension?

LinhWuerzburger commented on CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,52 @@
+## Contributing Guidelines
+
+The coding guidelines of ARTSS are mainly based on the [google coding guidelines](https://google.github.io/styleguide/cppguide.html). Below there are some rules which are not defined in the google standard or we want to especially emphasise them in order to achieve the most consistent code possible.
```
done in 3daa8cf

LinhWuerzburger commented on CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,52 @@
+## Contributing Guidelines
+
+The coding guidelines of ARTSS are mainly based on the [google coding guidelines](https://google.github.io/styleguide/cppguide.html). Below there are some rules which are not defined in the google standard or we want to especially emphasise them in order to achieve the most consistent code possible.
```
If I understand it correctly, you are planning to put every class into a namespace and not only the old static classes? And accordingly to add the rule to put everything in a namespace?

ChristianVonMach replied:
yeas every class should be in a namespace but of course not it's own. I my opinion we should use namespaces to represent the directories in our code like solver or pressure.

LinhWuerzburger replied:
fine by me. Could you open an issue regarding this discussion? I will think of a text for the guidelines.

LinhWuerzburger commented on CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,52 @@
+## Contributing Guidelines
+
+The coding guidelines of ARTSS are mainly based on the [google coding guidelines](https://google.github.io/styleguide/cppguide.html). Below there are some rules which are not defined in the google standard or we want to especially emphasise them in order to achieve the most consistent code possible.
```
Can you formulate a passage regarding these rules?

ChristianVonMach replied:
"ARTSS currently uses some static and global variables, but do not add new ones except for a exceptional good reason."

LinhWuerzburger commented on CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,52 @@
+## Contributing Guidelines
+
+The coding guidelines of ARTSS are mainly based on the [google coding guidelines](https://google.github.io/styleguide/cppguide.html). Below there are some rules which are not defined in the google standard or we want to especially emphasise them in order to achieve the most consistent code possible.
```
if adjusted how do you plan to adjust it or do you want to drop the rule?

ChristianVonMach replied:
I think if we go with the rule of 3-5-0 we are going to be okay.

LinhWuerzburger replied:
[3-5-0](https://google.github.io/styleguide/cppguide.html#Structs_vs._Tuples) doesn't seem the rule you are speaking of

LinhWuerzburger commented on CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,52 @@
+## Contributing Guidelines
+
+The coding guidelines of ARTSS are mainly based on the [google coding guidelines](https://google.github.io/styleguide/cppguide.html). Below there are some rules which are not defined in the google standard or we want to especially emphasise them in order to achieve the most consistent code possible.
```
I see the benefit in trailing return type syntax only for Lamba expressions. Therefore I propose to adjust the rule to 'for Lambda expressions the trailing return type syntax is used'. Or something similar.

LinhWuerzburger replied:
Is the trailing return type syntax something which can be used for our expressions in GlobalMacrosTypes?

ChristianVonMach replied:
trailing return types doesn't work on macros. we still have to relay on our compiler to deduce the correct type.

LinhWuerzburger commented:
I changed from relying on the google standard to our very own rules which are based on the google standard and in addition, I added that it is NOT necessary to know the google standard. Therefore we have to add every rule from the google standard which we want to use. If the rule is a large passage we can refer to the google standard for more detailed information.

LinhWuerzburger commented:
And I don't want to add the rule to not exceed 80 characters per line. This is just way too short. But we should add a section to line breaking. I don't mind to change the limit or to change the whole rule and to say that there won't be a line break at all. But I am not going with 80 characters.

LinhWuerzburger commented:
> And I don't want to add the rule to not exceed 80 characters per line. This is just way too short. But we should add a section to line breaking. I don't mind to change the limit or to change the whole rule and to say that there won't be a line break at all. But I am not going with 80 characters.

We could agree on a 100 character hard break while trying to take the 80 character line.
